fix duplicate stops issue

AvivSela opened the pull request:
the bug was that on every route that was created - a new set of stops was created too, without checking before of the stop is exist. so in case 3 routes share a stop the stop created 3 times. 

for fixing it - im using a dict of existing stops that validate im not creating a stop in case it already in it.

i think we should  discard all existing data and re-run for all past dates

OriHoch commented on open_bus_gtfs_etl/cli.py:
```diff
@@ -48,3 +48,7 @@ def load_to_db(**kwargs):
     api.cleanup_dated_paths(**kwargs)
+
+
+if __name__ == '__main__':
```
why did you add this? I removed it in this PR: https://github.com/hasadna/open-bus-gtfs-etl/pull/7/files#diff-02fcca231fb487d43a4a0e2f5eda40ce5be7e3087c678a13db9fbbdcfad30148L92

we should have a single entrypoint from the click CLI, otherwise it may cause code discrepencies

OriHoch commented on open_bus_gtfs_etl/gtfs_loader.py:
```diff
@@ -111,10 +132,15 @@ def same_route_exist_yesterday_adjust_dates(session: Session, route_from_gtfs):
 
 
+def _get_valid_stops_for_date(session: Session, date_to_analyze: date):
+    stops = session.query(Stop).filter(Stop.max_date <= date_to_analyze).order_by(Stop.max_date).all()
```
you are only getting older dates then date_to_analyze, won't this be a problem if it runs for a past date?
please make sure the code supports running for past dates too, not only for latest date

AvivSela replied:
I think that we need to talk about this slowly changing dimension more - There are some cases we both don't handle well. I suggest to separate the duplicate within an execution and duplicate between executions. the bug here was about duplicates within one execution - when two lines use same stop the stop was created twice.

in response to your example, the code will fetch the stop item with max date 20.10 and in case all fields are the same, no new record will be created - just the max date will be modified. 

i looked in the following SIRI etl logic and there you are fetching only items that belongs to the specific date. correct me if i'm wrong but there if we will use your use case - since no stop will be fetched - a new record will be created. 
```  
elif objname == 'stop':
            recorded_at_time = kwargs['recorded_at_time']
            return session.query(Stop).filter(
                Stop.min_date <= recorded_at_time.date(),
                recorded_at_time.date() <= Stop.max_date,
                Stop.is_from_gtfs == False,
                Stop.code == int(kwargs['stop_point_ref'])
            ).one_or_none()
```

OriHoch replied:
@AvivSela this bug is fixed?

AvivSela replied:
@OriHoch, yes, instead of using max date, I'm using min date , so stops that their min date was yesterday and max date will be tomorrow are included

OriHoch commented on open_bus_gtfs_etl/gtfs_stat/aggregations.py:
```diff
@@ -53,10 +53,11 @@ def trip_stats_aggregation(group):
+            for key in keys_for_all:
+                d[f'all_{key}'] = ';'.join([str(i) for i in group[key].tolist()])
+        except TypeError as e:
+            print(e)
```
please add a comment explaining why this exception can be ignored

OriHoch replied:
@AvivSela PR #13 was merged, please update this PR

OriHoch replied:
the try catch needs to be removed
also, PR #13 was merged, so you can update from main branch and remove this code

OriHoch commented on open_bus_gtfs_etl/gtfs_loader.py:
```diff
@@ -57,18 +58,38 @@ def _get_rides(self) -> List[Ride]:
                        self.all_trip_id_to_date.split(";"))]
 
-    def _get_stops(self) -> List[RouteStop]:
+    def _upsert_stop(self, stops, stop_code, stop_latlon, stop_name, stop_desc_city) -> Stop:
```
please add comments explaining what this code does

OriHoch commented on open_bus_gtfs_etl/gtfs_loader.py:
```diff
@@ -57,18 +58,38 @@ def _get_rides(self) -> List[Ride]:
+        return stop
+
+
+    def _get_stops(self, stops: Dict[int, Stop]) -> List[RouteStop]:
```
please add comments explaining what this code does

OriHoch commented:
@AvivSela any update on this?

AvivSela commented:
Hi, 

i changed the query condition so it also include stops with min date of yesterday and max date of tomorrow: 
https://github.com/hasadna/open-bus-gtfs-etl/blob/db2ce8eba83a89bed61e00fc0940d77ca15a0961/open_bus_gtfs_etl/gtfs_loader.py#L180

i changed the stop upsert logic:
https://github.com/hasadna/open-bus-gtfs-etl/blob/db2ce8eba83a89bed61e00fc0940d77ca15a0961/open_bus_gtfs_etl/gtfs_loader.py#L105

i added some unit-tests  for the stop upsert logic:
https://github.com/hasadna/open-bus-gtfs-etl/blob/db2ce8eba83a89bed61e00fc0940d77ca15a0961/tests/test_app.py#L192

i merged main into the branch - try-catch comment fixed

OriHoch commented on open_bus_gtfs_etl/gtfs_loader.py:
```diff
@@ -57,18 +58,57 @@ def _get_rides(self) -> List[Ride]:
+        stop = stops.get(int(stop_code))
+
+        if stop is None or stop.lat != float(lat) or stop.lon != float(lon) \
+                or stop.name != stop_name or stop.city != stop_desc_city:
```
@AvivSela please reply

AvivSela replied:
@OriHoch, yes. I'm fetching the prior or valid stop record for each stop and validate that new one will be created just in case stop is not exist, or it is different.

there are many cases i thought about. you can see them here: 
[def _upsert_stop(stops: Dict[int, Stop], stop_to_upsert: StopModel) -> Stop:](https://github.com/hasadna/open-bus-gtfs-etl/blob/16c313cd416a8d4a0ae8d9bbec035c862f51f7a4/open_bus_gtfs_etl/gtfs_loader.py#L105)

OriHoch commented:
following the discussion on hasadna/open-bus#359 - please make sure only to update data for the relevant date we are processing, don't assume the same data was relevant for intermediate dates too

OriHoch commented:
After more thought about it, let's not expand dates for now

just have a single record per day with min_date and max_date both set to the same date

Later we will can see how to implement this

OriHoch commented:
remove the min_date/max_date fields from the DB tables and have a single `date` field

AvivSela commented:
Hi, some descriptions for my new pushes: 
1. There was a change in the schema of the database and I changed the code accordingly.
2. The upsert mechanism was simplified due to our discussion and I convert it to simple logic where if exists I delete it and write it again. 
3. I added a test that simulates database using SQLite, and check that records are not written with duplications 
4. I worked with PyLint and fix some of the comments I got from there
5. known issue: there is some code that you wrote about calculating the stop time that I'm not so familiar with that should be modified due to the change in the schema.
6. known issue: after I tested it with real data, I saw in the gtfs_ride some strange values: there are multiple rides that belong to the same route with the same departure time but with different jurney_ref. I checked it and it is the real data came from GTFS Stat code that was written by Dan, we need to verify it is not "garbage in garbage out"
